### ISSUE TITLE
`redundant-expr`/`unreachable` false positive when narrowing type with generic to subtype

### ISSUE THREAD

**Issue by DetachHead** [KotlinIsland/basedmypy#694]
### Describe the problem, ie expected/actual result (if it's not blatantly obvious)

https://mypy-play.net/?mypy=basedmypy-latest&python=3.12&gist=36f3c4f2cdb3a0e91b0c831ec4a236e8

### Gist to reproduce

```python
from typing import Generic, _T as T
class Foo(Generic[T]): ...


class Bar(Foo[str]): ...


def asdf(foo: Foo[T]) -> None:
    if isinstance(foo, Bar):
        ...
```
